vm-virtio/src/vhost_user/net.rs

```diff
@@ -0,0 +1,309 @@
+            error!("vhost-user-net has queue number: {} larger than the max queue number: {} backend allowed\n",
+                vu_cfg.num_queues, max_queue_number);
+            return Err(Error::BadQueueNum);
+        }
+
+        let mut config_space = mac_addr.get_bytes().to_vec();
+        avail_features |= 1 << virtio_net::VIRTIO_NET_F_MAC;
+        let num_queue_pairs = (vu_cfg.num_queues / 2) as u16;
+        if (num_queue_pairs >= VIRTIO_NET_CTRL_MQ_VQ_PAIRS_MIN)
+            && (num_queue_pairs <= VIRTIO_NET_CTRL_MQ_VQ_PAIRS_MAX)
+        {
+            config_space.resize(CONFIG_SPACE_VUNET, 0);
+            let max_queue_pairs = num_queue_pairs.to_le_bytes();
+            config_space[CONFIG_SPACE_MAC + CONFIG_SPACE_STATUS..CONFIG_SPACE_VUNET]
+                .copy_from_slice(&max_queue_pairs);
+            avail_features |=
+                1 << virtio_net::VIRTIO_NET_F_CTRL_VQ | 1 << virtio_net::VIRTIO_NET_F_MQ;
+        }
+
+        vhost_user_net
+            .set_owner()
+            .map_err(Error::VhostUserSetOwner)?;
+
+        // Send set_vring_base here, since it could tell backends, like OVS + DPDK,
+        // how many virt queues to be handled, which backend required to know at early stage.
+        for i in 0..vu_cfg.num_queues {
+            vhost_user_net
+                .set_vring_base(i, 0)
+                .map_err(Error::VhostUserSetVringBase)?;
+        }
+
+        let queue_num = if (avail_features & (1 << virtio_net::VIRTIO_NET_F_CTRL_VQ) != 0)
+            && (avail_features & (1 << virtio_net::VIRTIO_NET_F_MQ) != 0)
+        {
+            1 + vu_cfg.num_queues
+        } else {
+            vu_cfg.num_queues
+        };
+
+        Ok(Net {
+            vhost_user_net,
+            kill_evt,
+            avail_features,
+            acked_features,
+            backend_features,
+            config_space,
+            queue_sizes: vec![vu_cfg.queue_size; queue_num],
+        })
+    }
+}
+
+impl Drop for Net {
+    fn drop(&mut self) {
+        if let Err(_e) = self.kill_evt.write(1) {
+            error!("failed to kill vhost-user-net with error {}", _e);
+        }
+    }
+}
+
+impl VirtioDevice for Net {
+    fn device_type(&self) -> u32 {
+        VirtioDeviceType::TYPE_NET as u32
+    }
+
+    fn queue_max_sizes(&self) -> &[u16] {
+        &self.queue_sizes
+    }
+
+    fn features(&self, page: u32) -> u32 {
+        match page {
+            0 => self.avail_features as u32,
+            1 => (self.avail_features >> 32) as u32,
+            _ => {
+                warn!("Received request for unknown features page: {}", page);
+                0u32
+            }
+        }
+    }
+
+    fn ack_features(&mut self, page: u32, value: u32) {
+        let mut v = match page {
+            0 => u64::from(value),
+            1 => u64::from(value) << 32,
+            _ => {
+                warn!("Cannot acknowledge unknown features page: {}", page);
+                0u64
+            }
+        };
```

Comment:
I'm not asking for any change, this is just a general comment here, unrelated to this PR: There's a lot of duplicated code between all virtio devices for the virtio features handling. Essentially, all devices implement the `features` and `ack_features` method in the same way.